src/librbd/io/ObjectRequest.cc

```diff
@@ -657,8 +675,137 @@ void ObjectWriteSameRequest::send_write() {
   AbstractObjectWriteRequest::send_write();
 }
 
+template <typename I>
+ObjectCmpextRequest<I>::ObjectCmpextRequest(ImageCtx *ictx, const std::string &oid,
+                     uint64_t object_no, uint64_t object_off, const ceph::bufferlist &cmp_bl,
+                     Context *completion)
+  : ObjectRequest<I>(util::get_image_ctx(ictx), oid, object_no, object_off, cmp_bl.length(), CEPH_NOSNAP,
+                  completion, false),
+  m_cmp_bl(cmp_bl), m_tried_parent(false), m_state(LIBRBD_AIO_CMPEXT_FLAT) {
+  guard_cmpext();
+}
+
+template <typename I>
+void ObjectCmpextRequest<I>::guard_cmpext()
+{
+  ImageCtx *image_ctx = this->m_ictx;
+  RWLock::RLocker snap_locker(image_ctx->snap_lock);
+  RWLock::RLocker parent_locker(image_ctx->parent_lock);
+
+  if (this->has_parent()) {
+    ldout(image_ctx->cct, 20) << __func__ << " guarding cmpext" << dendl;
+    m_state = LIBRBD_AIO_CMPEXT_GUARD;
+  }
+}
+
+
+template <typename I>
+bool ObjectCmpextRequest<I>::should_complete(int r)
+{
+  ImageCtx *image_ctx = this->m_ictx;
+  ldout(image_ctx->cct, 20) << "should_complete " << this << " "
+                            << this->m_oid << " "
+                            << this->m_object_off << "~" << this->m_object_len
+                            << " r = " << r << dendl;
+
+  bool finished = true;
+  switch (m_state) {
```

Comment:
I think you can actually just eliminate the need for a state enum. You first send it to the local image, if that returns -ENOENT, you haven't tried the parent, the parent exists, and their is an overlap, you send the request to the parent. Otherwise, you are done.